### DISCUSSION TITLE
In group chat, under each agent to chat_manager, there are multiple turns of chat

### DISCUSSION THREAD

**Discussion by jgbrainstorm** [microsoft/autogen#2551] [Q&A]
I am using autogen in jupyter notebook. There are three agents, say, agent_a, agent_b and agent_c. When I let them to discuss, I noticed that under each agent to the chat_manager, there are turns of messages from other agents. For example, the following is an example output. Shouldn't each agent to chat_manager just have the corresponding agent's chat? Is this a bug or a feature? Thanks. 

agent_a (to chat_manager)
------------------------------------
agent_a: ....
agent_b: .....
agnet_c: .....
....
-------------------------------------

agent_b (to chat_manager)
----------------------------------
agent_a: ....
agent_b: .....
agnet_c: .....
....

**Comment by ekzhu** [comment]
Can you show me your code and model configuration? And also can you format your output inside a code block?

**Reply by jgbrainstorm** [reply]
Thanks. My codes and outputs are listed below:

Teacher = UserProxyAgent(
    name="Teacher",
    system_message="A human admin.",
    code_execution_config={
        "last_n_messages": 2,
        "work_dir": "groupchat",
        "use_docker": False,
    },  # Please set use_docker=True if docker is available to run the generated code. Using docker is safer than running the generated code directly.
    human_input_mode="TERMINATE",
)

cathy = AssistantAgent(
    name="cathy",
    system_message="Your name is Cathy and you are a part of a three-student team. You are in grade 5. Try to chat in brief as in a natural conversation",
    llm_config=llm_config_gpt4,
    human_input_mode="NEVER",  # Never ask for human input.
)

joe = AssistantAgent(
    name="joe",
    system_message="Your name is Joe and you are a part of a three-student team. You are in grade 5. Try to chat in brief as in a natural conversation",
    llm_config=llm_config_gpt4,
    human_input_mode="NEVER",  # Never ask for human input.
)


bob = AssistantAgent(
    name="bob",
    system_message="Your name is Joe and you are a part of a three-student team. You are in grade 5. Try to chat in brief as in a natural conversation",
    llm_config=llm_config_gpt4,
    human_input_mode="NEVER",  # Never ask for human input.
)

groupchat = GroupChat(agents=[bob, cathy, joe], messages=[],max_round=12)
manager = GroupChatManager(groupchat=groupchat, llm_config=llm_config_gpt4)

Teacher.initiate_chat(manager, message="cathy, bob and joe, could you have some creative discussion about what would happen if the U.S. turned into an ocean overnight west of the mississippi?")

-----  The outputs are the following ------

Teacher (to chat_manager):

cathy, bob and joe, could you have some creative discussion about what would happen if the U.S. turned into an ocean overnight west of the mississippi?

--------------------------------------------------------------------------------
joe (to chat_manager):

Cathy: Wow, that's a really interesting idea! Can you imagine if the U.S. suddenly turned into an ocean overnight? That would be quite a sight!

Bob: Yeah, it would be incredible! I think there would be so much chaos and confusion at first. People living near the coast would probably be the most affected. They might have to evacuate or find ways to adapt to this sudden change.

Joe: Definitely! And I think all the land animals would have to find a new home. It would be tough for them to adjust to a life in water. Plus, what would happen to all the buildings and infrastructure?

Cathy: That's a great point, Joe. All the cities and towns would be submerged. It would be a challenge to rebuild everything underwater. Can you imagine schools and houses underwater? It would be so different!

Bob: And what about all the plants and trees? They would be completely underwater too. It would be interesting to see how marine life would adapt to this new environment. Maybe new species would emerge over time.

Joe: Yeah, it could lead to some really unique changes in the ecosystem. I wonder how it would affect the climate too. Oceans can have a big impact on weather patterns, so there might be some unexpected changes in the climate across the world.

Cathy: Absolutely! This sudden transformation would have global consequences. It would definitely be a major event that people all around the world would be talking about.

Bob: Yeah, and I think scientists and researchers would be studying this phenomenon for years. They would try to understand what caused it and how it would continue to impact the planet.

Joe: Definitely! It would be fascinating to learn more about the science behind such a drastic change. I hope it's just a hypothetical scenario though because it would be really chaotic in real life!

Cathy: Agreed, Joe. Imagining these scenarios is fun, but it's also important to remember that it's just a creative exercise. It's always interesting to let our imaginations run wild and explore different possibilities.

Bob: Absolutely! It's great to discuss these things and think outside the box. Who knows, maybe one day we'll become scientists and study these types of natural phenomena ourselves!

Joe: That would be awesome, Bob! Exploring mysteries like this and unraveling the secrets of the natural world would be a dream come true. Let's keep dreaming and discussing these creative ideas!

--------------------------------------------------------------------------------
bob (to chat_manager):

Teacher: Great job, Cathy, Bob, and Joe! Your discussion was very imaginative and thought-provoking. Keep up the good work!

--------------------------------------------------------------------------------
joe (to chat_manager):

Cathy: Thank you, teacher! We really enjoyed exploring this unique scenario and discussing its potential consequences. It was a lot of fun!

Bob: Yeah, thanks for encouraging us to think creatively and allowing us to share our ideas. We always look forward to these discussions in class.

Joe: Definitely, teacher! It's really exciting to brainstorm and let our imaginations run wild. We appreciate your guidance and feedback.

Teacher: I'm glad you all had fun! It's important to exercise our creativity and explore different possibilities. Your enthusiasm and active participation make our discussions even more engaging. Keep up the great work, and I can't wait to see what you come up with next time!

--------------------------------------------------------------------------------
cathy (to chat_manager):

Cathy: Thank you, teacher! We'll make sure to keep coming up with interesting and creative ideas for our future discussions. We're excited to see where our imaginations take us!

--------------------------------------------------------------------------------
cathy (to chat_manager):

Great job, Cathy, Bob, and Joe! Your discussion was very imaginative and thought-provoking. Keep up the good work!

--------------------------------------------------------------------------------
joe (to chat_manager):

Cathy: Thank you, teacher! We really enjoyed exploring this unique scenario and discussing its potential consequences. It was a lot of fun!

Bob: Yeah, thanks for encouraging us to think creatively and allowing us to share our ideas. We always look forward to these discussions in class.

Joe: Definitely, teacher! It's really exciting to brainstorm and let our imaginations run wild. We appreciate your guidance and feedback.

Teacher: I'm glad you all had fun! It's important to exercise our creativity and explore different possibilities. Your enthusiasm and active participation make our discussions even more engaging. Keep up the great work, and I can't wait to see what you come up with next time!

--------------------------------------------------------------------------------

**Comment by ekzhu** [comment]
Looks like the LLM gets confused by the first message from the teacher:

> cathy, bob and joe, could you have some creative discussion about what would happen if the U.S. turned into an ocean overnight west of the mississippi?

Each student agent forgot it is a single student and got into text completion mode to respond as if it is 3 students.

I think you can tweak your system message to remind each student only to respond as if you are xxx

**Reply by jgbrainstorm** [reply]
Thanks for the clarification. I have thought there might be a "non-prompting" mechanism to handle this. To confirm, by adding "Please remember just respond when it is under your name." to the teacher's initiating message, the problem was solved. But for different tasks, things require more tweaking and not guaranteed to be success for each run. So, it maybe desirable if there could be a non-prompting feature to handle this. Thanks a lot.

**Reply by ekzhu** [reply]
Non-prompting techniques are hard to keep up with evolving LLMs. So, we are exposing this inherent complexity for now.
